File: src/main/java/algo/breadthFirstSearch/OpenTheLock_752.java
package algo.breadthFirstSearch;

import java.util.*;

public class OpenTheLock_752 {
  public static int openLock(String[] deadends, String target) {
    var start = "0000";
    var rounds = 0;
    var dead = new HashSet<>(Arrays.asList(deadends));
    var visited = new HashSet<String>();

    if (dead.contains(start)) {
      return -1;
    }

    var q = new ArrayDeque<String>();
    q.offer(start);
    while (!q.isEmpty()) {
      var size = q.size();
      var list = new ArrayList<String>();
      for (var i = 0; i < size; i++) {
        var next = q.poll();
        if (target.equals(next)) {
          return rounds;
        }
        if (!visited.add(next)) {
          continue;
        }
        list.addAll(getNext(next, dead));
      }
      list.forEach(q::offer);
      rounds++;
    }
    return -1;
  }

  private static List<String> getNext(String s, Set<String> dead) {
    var result = new ArrayList<String>();
    for (var i = 0; i < s.length(); i++) {
      String res1 = "", res2 = "";
      for (var j = 0; j < s.length(); j++) {
        if (j == i) {
          var c1 = s.charAt(i) == '9' ? '0' : (char)(s.charAt(i) + 1);
          var c2 = s.charAt(i) == '0' ? '9' : (char)(s.charAt(i) - 1);
          res1 += c1;
          res2 += c2;
        } else {
          res1 += s.charAt(j);
          res2 += s.charAt(j);
        }
      }
      for (var res : new String[]{res1, res2}) {
        if (!dead.contains(res)) {
          result.add(res);
        }
      }
    }
    return result;
  }

  public static void main(String[] args) {
    System.out.println(openLock(new String[]{"0201","0101","0102","1212","2002"}, "0202"));
    System.out.println(openLock(new String[]{"8887","8889","8878","8898","8788","8988","7888","9888"}, "8888"));
  }
}

File: src/main/kotlin/algo/breadthFirstSearch/OpenTheLock_752.kt
package algo.breadthFirstSearch

import java.lang.StringBuilder
import java.util.LinkedList

class KtOpenTheLock_752 {
    fun openLock(deadends: Array<String>, target: String): Int {
        fun nextStates(state: String, deadends: Set<String>): Set<String> {
            val states: MutableSet<String> = mutableSetOf()
            for (i in state.indices) {
                when (state[i]) {
                    '0' -> {
                        val sb1 = StringBuilder(state)
                        sb1.setCharAt(i, '1')
                        states.add(sb1.toString())
                        sb1.setCharAt(i, '9')
                        states.add(sb1.toString())
                    }
                    '9' -> {
                        val sb1 = StringBuilder(state)
                        sb1.setCharAt(i, '0')
                        states.add(sb1.toString())
                        sb1.setCharAt(i, '8')
                        states.add(sb1.toString())
                    }
                    else -> {
                        val sb1 = StringBuilder(state)
                        sb1.setCharAt(i, state[i] - 1)
                        states.add(sb1.toString())
                        sb1.setCharAt(i, state[i] + 1)
                        states.add(sb1.toString())
                    }
                }
            }
            return states.filter { !deadends.contains(it) }.toSet()
        }

        if (target == "0000") {
            return 0
        }
        if (deadends.contains(target)) {
            return -1
        }

        var steps = 1
        val visited: MutableSet<String> = mutableSetOf()
        val q = LinkedList<String>()
        q.offer("0000")
        while (q.isNotEmpty()) {
            val size = q.size
            for (i in 1..size) {
                val e = q.poll()
                if (!visited.add(e)) {
                    continue
                }
                if (deadends.contains(e)) {
                    continue
                }
                val next = nextStates(e, deadends.toSet())
                if (next.contains(target)) {
                    return steps
                }
                next.filter { !visited.contains(it) }.forEach { q.offer(it) }
            }
            steps++
        }

        return -1
    }
}

fun main() {
    val openLock = KtOpenTheLock_752()
    println(openLock.openLock(arrayOf("0201","0101","0102","1212","2002"), "0202"))
    println(openLock.openLock(arrayOf("8888"), "0009"))
    println(openLock.openLock(arrayOf("8887","8889","8878","8898","8788","8988","7888","9888"), "8888"))
}